File: Computational_Physics_2/1.Lectures/8.Computo_en_paralelo_OPENMP/examples/example_1_Hello_World/Fortran/example_7_omp_dim_array.c
// https://docs-dev.nersc.gov/cgpu/examples/
// compiled with the GCC Fortran compiler gfortran from the gcc/8.1.1-openacc-gcc-8-branch-20190215 module:
// $ clang -fopenmp -fopenmp-targets=nvptx64-nvidia-cuda -o main.ex main.c
// $ srun -n 1 nvprof ./main.ex

#include <stdio.h>

#define MAX 32768

int main()
{
  int x[MAX], i;
  for (i = 0; i < MAX; ++i) {
    x[i] = 42;
  }

  #pragma omp target teams distribute parallel for simd map(tofrom: x[0:MAX])
  for (i = 0; i < MAX; ++i) {
    x[i] = x[i] + 1;
  }
  printf("After the target region is executed, x[0] = %d\n", x[0]);
  return 0;
}

File: Computational_Physics_2/1.Lectures/8.Computo_en_paralelo_OPENMP/examples/example_1_Hello_World/Fortran/example_7_omp_dim_array.f90
! https://docs-dev.nersc.gov/cgpu/examples/
! compiled with the GCC Fortran compiler gfortran from the gcc/8.1.1-openacc-gcc-8-branch-20190215 module:
! $ gfortran -fopenmp -foffload=nvptx-none="-Ofast -lm -misa=sm_35" -o main.ex main.f90
! $ srun -n 1 nvprof ./main.ex


program main
  implicit none

  integer, parameter :: sz = 32768
  integer, dimension(sz) :: arr
  integer :: i

  do i = 1, sz
    arr(i) = 42
  end do

  !$omp target teams distribute parallel do map(tofrom: arr(1:sz))
  do i = 1, sz
    arr(i) = arr(i) + 1
  end do

  print *, "After the target region is executed, arr(1) = ", arr(1)

end program main
